### ISSUE TITLE
lint unnecessary literal casts

### ISSUE THREAD

**Issue by euclio** [rust-lang/rust-clippy#6116]
### What it does

Lints unnecessary as-casts of literals when they could be written using literal syntax.

### Categories (optional)

- Kind: `clippy::style`

*What is the advantage of the recommended code over the original code*

- Avoids accidental truncation
- Uses syntax dedicated to this purpose

### Drawbacks

None.

### Example

```rust
1 as u32
```

Could be written as:

```rust
1u32
```

**Comment by leonardo-m** [comment]
Better to suggest: 1_u32

**Comment by ebroto** [comment]
This would be currently linted by [`as_conversions`](https://rust-lang.github.io/rust-clippy/master/#as_conversions) ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=34e652df7a7dab2188a9923f0965ed90)). The docs imply it's allow-by-default because it has false positives. 

The lint could be improved though to generate suggestions as the one you gave.

**Comment by euclio** [comment]
I think the lints serve different purposes. I `1 as u32` is almost certainly a mistake, while there are legitimate uses of as casts in general (which is why it's allow-by-default).

**Comment by flip1995** [comment]
Yes `as_conversions` is a restriction lint, with the design that it lints **all** `as` conversions, regardless what it converts.

**Comment by geoffreycopin** [comment]
Hi, I'd be glad to give this a shot !

**Comment by geoffreycopin** [comment]
When testing my implementation, I noticed that there is some overlap with a [https://github.com/rust-lang/rust-clippy/blob/master/tests/ui/unnecessary_cast_fixable.rs](previously existing_lint) that flags int literals that are casted to a float type.

Would it be suitable to just extend the aforementioned lint so that it become "casting number literal to <cast_to> is unnecessary" ? If not, what would be a good way to handle the overlap ?

**Comment by ebroto** [comment]
@geoffreycopin I think what you suggest makes sense as that lint was already extended in #3842 to add the float cases.

Even if later there's a decision to not enhance that lint, your code should go in that module to be able to handle the overlap.

**Comment by ebroto** [comment]
Fixed by #6187
